File: server/routers/beads_api.py
"""
Beads API Router
================

Host-based API wrapper for beads commands. Agents call these endpoints instead
of running `bd` directly. The host runs `bd` commands on the project directory.

This provides:
- Centralized beads access (no direct bd in containers)
- Concurrency control via per-project locks
- Consistent JSON responses
"""

import asyncio
import json
import logging
import re
import subprocess
import sys
from pathlib import Path
from typing import Any

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

logger = logging.getLogger(__name__)

# Per-project locks for write operations
_beads_locks: dict[str, asyncio.Lock] = {}


def _get_project_path(project_name: str) -> Path:
    """Get project path from registry."""
    _root = Path(__file__).parent.parent.parent
    if str(_root) not in sys.path:
        sys.path.insert(0, str(_root))

    from registry import get_project_path
    return get_project_path(project_name)


def validate_project_name(name: str) -> str:
    """Validate and sanitize project name to prevent path traversal."""
    if not re.match(r'^[a-zA-Z0-9_-]{1,50}$', name):
        raise HTTPException(status_code=400, detail="Invalid project name")
    return name


def validate_issue_id(issue_id: str) -> str:
    """Validate issue ID format."""
    # Allow formats like: beads-1, feat-42, project-abc123
    if not re.match(r'^[a-zA-Z]+-[a-zA-Z0-9]+$', issue_id):
        raise HTTPException(status_code=400, detail="Invalid issue ID format")
    return issue_id


async def run_beads_command(project_name: str, args: list[str]) -> dict[str, Any]:
    """
    Run bd command in project directory on host.

    Args:
        project_name: Name of the project
        args: Command arguments (e.g., ["list", "--json"])

    Returns:
        Parsed JSON output or error dict
    """
    project_path = _get_project_path(project_name)

    if not project_path:
        return {"error": f"Project '{project_name}' not found in registry"}

    if not project_path.exists():
        return {"error": f"Project directory not found: {project_path}"}

    try:
        result = await asyncio.to_thread(
            subprocess.run,
            ["bd", "--no-daemon", *args],
            cwd=project_path,
            capture_output=True,
            text=True,
            timeout=60,
        )

        if result.returncode != 0:
            error_msg = result.stderr.strip() if result.stderr else f"Command failed with exit code {result.returncode}"
            logger.warning(f"bd command failed: bd {' '.join(args)} - {error_msg}")
            return {"error": error_msg}

        # Try to parse JSON output
        stdout = result.stdout.strip()
        if not stdout:
            return {"success": True, "data": []}

        try:
            return {"success": True, "data": json.loads(stdout)}
        except json.JSONDecodeError:
            # Some commands return plain text
            return {"success": True, "output": stdout}

    except subprocess.TimeoutExpired:
        return {"error": "Command timed out"}
    except FileNotFoundError:
        return {"error": "bd command not found. Is beads installed?"}
    except Exception as e:
        logger.exception(f"Error running beads command: {e}")
        return {"error": str(e)}


async def run_beads_write_command(project_name: str, args: list[str]) -> dict[str, Any]:
    """
    Run a write command with project-level locking.

    Write operations (create, update, close, reopen, sync) are serialized
    per-project to avoid race conditions.
    """
    lock = _beads_locks.setdefault(project_name, asyncio.Lock())

    async with lock:
        return await run_beads_command(project_name, args)


# =============================================================================
# Request/Response Models
# =============================================================================

class IssueCreate(BaseModel):
    """Request model for creating an issue."""
    title: str = Field(..., min_length=1, max_length=200)
    description: str = Field(default="")
    type: str = Field(default="task")  # task, bug, feature, epic
    priority: int = Field(default=2, ge=0, le=4)  # 0=P0 (critical) to 4=P4 (backlog)
    labels: list[str] = Field(default_factory=list)


class IssueUpdate(BaseModel):
    """Request model for updating an issue."""
    title: str | None = None
    description: str | None = None
    status: str | None = None  # open, in_progress, closed
    priority: int | None = Field(default=None, ge=0, le=4)
    assignee: str | None = None


class IssueClose(BaseModel):
    """Request model for closing an issue."""
    reason: str | None = None


# =============================================================================
# Router
# =============================================================================

router = APIRouter(prefix="/api/projects/{project_name}/beads", tags=["beads"])


@router.get("/list")
async def list_issues(project_name: str, status: str | None = None):
    """
    List all issues for a project.

    Query params:
        status: Filter by status (open, in_progress, closed)
    """
    project_name = validate_project_name(project_name)

    args = ["list", "--json"]
    if status:
        args.extend(["--status", status])

    result = await run_beads_command(project_name, args)

    if "error" in result:
        raise HTTPException(status_code=500, detail=result["error"])

    return result.get("data", [])


@router.get("/ready")
async def ready_issues(project_name: str):
    """
    List issues ready for work (no blockers).

    Returns issues that are open and have no blocking dependencies.
    """
    project_name = validate_project_name(project_name)

    result = await run_beads_command(project_name, ["ready", "--json"])

    if "error" in result:
        raise HTTPException(status_code=500, detail=result["error"])

    return result.get("data", [])


@router.get("/show/{issue_id}")
async def show_issue(project_name: str, issue_id: str):
    """
    Get details of a specific issue.
    """
    project_name = validate_project_name(project_name)
    issue_id = validate_issue_id(issue_id)

    result = await run_beads_command(project_name, ["show", issue_id, "--json"])

    if "error" in result:
        if "not found" in result["error"].lower():
            raise HTTPException(status_code=404, detail=f"Issue {issue_id} not found")
        raise HTTPException(status_code=500, detail=result["error"])

    data = result.get("data", [])
    if isinstance(data, list) and data:
        return data[0]
    return data


@router.get("/stats")
async def stats(project_name: str):
    """
    Get project statistics (open/closed/blocked counts).
    """
    project_name = validate_project_name(project_name)

    result = await run_beads_command(project_name, ["stats", "--json"])

    if "error" in result:
        raise HTTPException(status_code=500, detail=result["error"])

    return result.get("data", result.get("output", {}))


@router.post("/create")
async def create_issue(project_name: str, issue: IssueCreate):
    """
    Create a new issue.
    """
    project_name = validate_project_name(project_name)

    # Build command args
    args = [
        "create",
        "--title", issue.title,
        "--type", issue.type,
        "--priority", f"P{issue.priority}",
        "--json",
    ]

    if issue.description:
        args.extend(["--description", issue.description])

    if issue.labels:
        args.extend(["--labels", ",".join(issue.labels)])

    result = await run_beads_write_command(project_name, args)

    if "error" in result:
        raise HTTPException(status_code=500, detail=result["error"])

    return result.get("data", {"success": True})


@router.patch("/update/{issue_id}")
async def update_issue(project_name: str, issue_id: str, update: IssueUpdate):
    """
    Update an issue's fields.

    Only provided fields will be updated.
    """
    project_name = validate_project_name(project_name)
    issue_id = validate_issue_id(issue_id)

    # Build command args
    args = ["update", issue_id]

    if update.title is not None:
        args.extend(["--title", update.title])

    if update.description is not None:
        args.extend(["--description", update.description])

    if update.status is not None:
        args.extend(["--status", update.status])

    if update.priority is not None:
        args.extend(["--priority", f"P{update.priority}"])

    if update.assignee is not None:
        args.extend(["--assignee", update.assignee])

    # Must have at least one update
    if len(args) == 2:
        raise HTTPException(status_code=400, detail="No update fields provided")

    result = await run_beads_write_command(project_name, args)

    if "error" in result:
        if "not found" in result["error"].lower():
            raise HTTPException(status_code=404, detail=f"Issue {issue_id} not found")
        raise HTTPException(status_code=500, detail=result["error"])

    return {"success": True, "message": f"Issue {issue_id} updated"}


@router.post("/close/{issue_id}")
async def close_issue(project_name: str, issue_id: str, body: IssueClose | None = None):
    """
    Close an issue.
    """
    project_name = validate_project_name(project_name)
    issue_id = validate_issue_id(issue_id)

    args = ["close", issue_id]

    if body and body.reason:
        args.extend(["--reason", body.reason])

    result = await run_beads_write_command(project_name, args)

    if "error" in result:
        if "not found" in result["error"].lower():
            raise HTTPException(status_code=404, detail=f"Issue {issue_id} not found")
        raise HTTPException(status_code=500, detail=result["error"])

    return {"success": True, "message": f"Issue {issue_id} closed"}


@router.post("/reopen/{issue_id}")
async def reopen_issue(project_name: str, issue_id: str):
    """
    Reopen a closed issue.
    """
    project_name = validate_project_name(project_name)
    issue_id = validate_issue_id(issue_id)

    result = await run_beads_write_command(project_name, ["reopen", issue_id])

    if "error" in result:
        if "not found" in result["error"].lower():
            raise HTTPException(status_code=404, detail=f"Issue {issue_id} not found")
        raise HTTPException(status_code=500, detail=result["error"])

    return {"success": True, "message": f"Issue {issue_id} reopened"}


@router.post("/sync")
async def sync_issues(project_name: str):
    """
    Sync beads with git remote.
    """
    project_name = validate_project_name(project_name)

    result = await run_beads_write_command(project_name, ["sync"])

    if "error" in result:
        raise HTTPException(status_code=500, detail=result["error"])

    return {"success": True, "message": "Beads synced with remote"}


@router.post("/dep/add")
async def add_dependency(project_name: str, issue_id: str, depends_on: str):
    """
    Add a dependency between issues.

    Args:
        issue_id: The issue that depends on another
        depends_on: The issue that must be completed first
    """
    project_name = validate_project_name(project_name)
    issue_id = validate_issue_id(issue_id)
    depends_on = validate_issue_id(depends_on)

    result = await run_beads_write_command(
        project_name, ["dep", "add", issue_id, depends_on]
    )

    if "error" in result:
        raise HTTPException(status_code=500, detail=result["error"])

    return {"success": True, "message": f"Added dependency: {issue_id} depends on {depends_on}"}


class CommentAdd(BaseModel):
    """Request model for adding a comment."""
    comment: str = Field(..., min_length=1)


@router.post("/comments/{issue_id}")
async def add_comment(project_name: str, issue_id: str, body: CommentAdd):
    """
    Add a comment to an issue.
    """
    project_name = validate_project_name(project_name)
    issue_id = validate_issue_id(issue_id)

    result = await run_beads_write_command(
        project_name, ["comments", issue_id, "--add", body.comment]
    )

    if "error" in result:
        if "not found" in result["error"].lower():
            raise HTTPException(status_code=404, detail=f"Issue {issue_id} not found")
        raise HTTPException(status_code=500, detail=result["error"])

    return {"success": True, "message": f"Comment added to {issue_id}"}

File: packages/server/src/routers/beads-api.ts
/**
 * Beads API Router
 * ================
 *
 * Host-based API wrapper for beads commands. Agents call these endpoints instead
 * of running `bd` directly. The host runs `bd` commands on the project directory.
 *
 * This provides:
 * - Centralized beads access (no direct bd in containers)
 * - Concurrency control via unified BeadsManager
 * - Consistent JSON responses
 */

import { Hono } from 'hono';
import { HTTPException } from 'hono/http-exception';
import { z } from 'zod';

import {
  getBeadsManager,
  type BeadsTask,
} from '../services/beads-manager.js';
import {
  getContainer,
  updateContainerStatus,
  setLastClosedFeature,
} from '../db/crud.js';

// =============================================================================
// Validation Helpers
// =============================================================================

/**
 * Validate and sanitize project name to prevent path traversal.
 */
function validateProjectName(name: string): string {
  if (!/^[a-zA-Z0-9_-]{1,50}$/.test(name)) {
    throw new HTTPException(400, { message: 'Invalid project name' });
  }
  return name;
}

/**
 * Validate issue ID format.
 */
function validateIssueId(issueId: string): string {
  // Allow formats like: beads-1, feat-42, project-abc123
  if (!/^[a-zA-Z]+-[a-zA-Z0-9]+$/.test(issueId)) {
    throw new HTTPException(400, { message: 'Invalid issue ID format' });
  }
  return issueId;
}

// =============================================================================
// Request/Response Schemas
// =============================================================================

const IssueCreateSchema = z.object({
  title: z.string().min(1).max(200),
  description: z.string().default(''),
  type: z.string().default('task'), // task, bug, feature, epic
  priority: z.number().int().min(0).max(4).default(2), // 0=P0 (critical) to 4=P4 (backlog)
  labels: z.array(z.string()).default([]),
});

const IssueUpdateSchema = z.object({
  title: z.string().optional(),
  description: z.string().optional(),
  status: z.string().optional(), // open, in_progress, closed
  priority: z.number().int().min(0).max(4).optional(),
  assignee: z.string().optional(),
});

const IssueCloseSchema = z.object({
  reason: z.string().optional(),
});

const DependencyAddSchema = z.object({
  issue_id: z.string(),
  depends_on: z.string(),
});

const CommentAddSchema = z.object({
  comment: z.string().min(1),
});

// =============================================================================
// Helper Functions
// =============================================================================

/**
 * Parse and validate JSON body with a Zod schema.
 */
async function parseBody<T extends z.ZodType>(
  c: { req: { json: () => Promise<unknown> } },
  schema: T
): Promise<z.infer<T>> {
  const body = await c.req.json();
  const result = schema.safeParse(body);
  if (!result.success) {
    throw new HTTPException(400, { message: `Validation error: ${result.error.message}` });
  }
  return result.data;
}

/**
 * Get container number from X-Container-Number header.
 */
function getContainerNumber(c: { req: { header: (name: string) => string | undefined } }): number | null {
  const header = c.req.header('X-Container-Number');
  if (header) {
    const num = parseInt(header, 10);
    if (!isNaN(num)) {
      return num;
    }
  }
  return null;
}

// =============================================================================
// Router Setup
// =============================================================================

const beadsApiRouter = new Hono();

// =============================================================================
// Route Handlers
// =============================================================================

// GET /api/projects/:name/beads/list - List all issues
beadsApiRouter.get('/:name/beads/list', async (c) => {
  const projectName = validateProjectName(c.req.param('name'));
  const status = c.req.query('status');

  try {
    const manager = await getBeadsManager(projectName);
    let tasks: BeadsTask[];

    if (status) {
      tasks = manager.getTasksByStatus(status);
    } else {
      tasks = manager.getTasks();
    }

    return c.json(tasks);
  } catch (error) {
    const err = error as Error;
    throw new HTTPException(500, { message: err.message });
  }
});

// GET /api/projects/:name/beads/ready - List issues ready for work
beadsApiRouter.get('/:name/beads/ready', async (c) => {
  const projectName = validateProjectName(c.req.param('name'));

  try {
    const manager = await getBeadsManager(projectName);
    const result = await manager.runReadCommand(['ready', '--json']);

    if (result.error) {
      throw new HTTPException(500, { message: result.error });
    }

    return c.json(result.data ?? []);
  } catch (error) {
    if (error instanceof HTTPException) throw error;
    const err = error as Error;
    throw new HTTPException(500, { message: err.message });
  }
});

// GET /api/projects/:name/beads/show/:issueId - Get issue details
beadsApiRouter.get('/:name/beads/show/:issueId', async (c) => {
  const projectName = validateProjectName(c.req.param('name'));
  const issueId = validateIssueId(c.req.param('issueId'));

  try {
    const manager = await getBeadsManager(projectName);
    const result = await manager.runReadCommand(['show', issueId, '--json']);

    if (result.error) {
      if (result.error.toLowerCase().includes('not found')) {
        throw new HTTPException(404, { message: `Issue ${issueId} not found` });
      }
      throw new HTTPException(500, { message: result.error });
    }

    const data = result.data;
    if (Array.isArray(data) && data.length > 0) {
      return c.json(data[0]);
    }
    return c.json(data);
  } catch (error) {
    if (error instanceof HTTPException) throw error;
    const err = error as Error;
    throw new HTTPException(500, { message: err.message });
  }
});

// GET /api/projects/:name/beads/stats - Get project statistics
beadsApiRouter.get('/:name/beads/stats', async (c) => {
  const projectName = validateProjectName(c.req.param('name'));

  try {
    const manager = await getBeadsManager(projectName);
    const stats = manager.getStats();
    return c.json(stats);
  } catch (error) {
    const err = error as Error;
    throw new HTTPException(500, { message: err.message });
  }
});

// POST /api/projects/:name/beads/create - Create a new issue
beadsApiRouter.post('/:name/beads/create', async (c) => {
  const projectName = validateProjectName(c.req.param('name'));
  const issue = await parseBody(c, IssueCreateSchema);

  try {
    const manager = await getBeadsManager(projectName);
    const result = await manager.createIssue(
      issue.title,
      issue.type,
      issue.priority,
      issue.description,
      issue.labels.length > 0 ? issue.labels : undefined
    );

    if (result.error) {
      throw new HTTPException(500, { message: result.error });
    }

    return c.json(result.data ?? { success: true });
  } catch (error) {
    if (error instanceof HTTPException) throw error;
    const err = error as Error;
    throw new HTTPException(500, { message: err.message });
  }
});

// POST /api/projects/:name/beads/claim - Atomically claim the next available issue
beadsApiRouter.post('/:name/beads/claim', async (c) => {
  const projectName = validateProjectName(c.req.param('name'));
  const containerNumber = getContainerNumber(c);

  try {
    const manager = await getBeadsManager(projectName);

    // Get list of ready issues (open, no blockers)
    const readyResult = await manager.runReadCommand(['ready', '--json']);
    if (readyResult.error) {
      throw new HTTPException(500, { message: readyResult.error });
    }

    const allIssues = (readyResult.data ?? []) as BeadsTask[];
    // Filter to only open issues (bd ready includes in_progress)
    const issues = allIssues.filter((i) => i.status === 'open');

    if (issues.length === 0) {
      return c.json({ success: false, message: 'No issues available to claim', issue: null });
    }

    // Get the first available open issue
    const issue = issues[0]!;
    const issueId = issue.id;

    if (!issueId) {
      throw new HTTPException(500, { message: 'Issue missing ID' });
    }

    // Claim it - update to in_progress
    const updateResult = await manager.updateIssue(issueId, { status: 'in_progress' });

    if (updateResult.error) {
      throw new HTTPException(500, { message: updateResult.error });
    }

    // Update container's current_feature in DB
    if (containerNumber !== null) {
      try {
        updateContainerStatus(projectName, containerNumber, 'coding', { currentFeature: issueId });
        console.debug(`Set current_feature=${issueId} for container ${containerNumber}`);
      } catch (e) {
        console.warn(`Failed to update container current_feature: ${e}`);
        // Don't fail the claim for tracking errors
      }
    }

    return c.json({
      success: true,
      message: `Claimed issue ${issueId}`,
      issue,
    });
  } catch (error) {
    if (error instanceof HTTPException) throw error;
    const err = error as Error;
    throw new HTTPException(500, { message: err.message });
  }
});

// PATCH /api/projects/:name/beads/update/:issueId - Update an issue
beadsApiRouter.patch('/:name/beads/update/:issueId', async (c) => {
  const projectName = validateProjectName(c.req.param('name'));
  const issueId = validateIssueId(c.req.param('issueId'));
  const update = await parseBody(c, IssueUpdateSchema);

  // Must have at least one update field
  if (
    update.title === undefined &&
    update.description === undefined &&
    update.status === undefined &&
    update.priority === undefined &&
    update.assignee === undefined
  ) {
    throw new HTTPException(400, { message: 'No update fields provided' });
  }

  try {
    const manager = await getBeadsManager(projectName);
    const result = await manager.updateIssue(issueId, update);

    if (result.error) {
      if (result.error.toLowerCase().includes('not found')) {
        throw new HTTPException(404, { message: `Issue ${issueId} not found` });
      }
      throw new HTTPException(500, { message: result.error });
    }

    return c.json({ success: true, message: `Issue ${issueId} updated` });
  } catch (error) {
    if (error instanceof HTTPException) throw error;
    const err = error as Error;
    throw new HTTPException(500, { message: err.message });
  }
});

// POST /api/projects/:name/beads/close/:issueId - Close an issue
beadsApiRouter.post('/:name/beads/close/:issueId', async (c) => {
  const projectName = validateProjectName(c.req.param('name'));
  const issueId = validateIssueId(c.req.param('issueId'));
  const containerNumber = getContainerNumber(c);

  let body: z.infer<typeof IssueCloseSchema> | null = null;
  try {
    body = await parseBody(c, IssueCloseSchema);
  } catch {
    // Body is optional for close
  }

  try {
    const manager = await getBeadsManager(projectName);
    const result = await manager.closeIssue(issueId, body?.reason);

    if (result.error) {
      if (result.error.toLowerCase().includes('not found')) {
        throw new HTTPException(404, { message: `Issue ${issueId} not found` });
      }
      throw new HTTPException(500, { message: result.error });
    }

    // Track which container closed this feature
    if (containerNumber !== null) {
      try {
        setLastClosedFeature(projectName, containerNumber, issueId);
        console.debug(`Tracked closed feature ${issueId} for container ${containerNumber}`);

        // Clear current_feature if it matches the closed issue
        const container = getContainer(projectName, containerNumber, 'coding');
        if (container && container.currentFeature === issueId) {
          updateContainerStatus(projectName, containerNumber, 'coding', { currentFeature: '' });
          console.debug(`Cleared current_feature for container ${containerNumber}`);
        }
      } catch (e) {
        console.warn(`Failed to track closed feature: ${e}`);
        // Don't fail the close operation for tracking errors
      }
    }

    return c.json({ success: true, message: `Issue ${issueId} closed` });
  } catch (error) {
    if (error instanceof HTTPException) throw error;
    const err = error as Error;
    throw new HTTPException(500, { message: err.message });
  }
});

// POST /api/projects/:name/beads/reopen/:issueId - Reopen an issue
beadsApiRouter.post('/:name/beads/reopen/:issueId', async (c) => {
  const projectName = validateProjectName(c.req.param('name'));
  const issueId = validateIssueId(c.req.param('issueId'));

  try {
    const manager = await getBeadsManager(projectName);
    const result = await manager.reopenIssue(issueId);

    if (result.error) {
      if (result.error.toLowerCase().includes('not found')) {
        throw new HTTPException(404, { message: `Issue ${issueId} not found` });
      }
      throw new HTTPException(500, { message: result.error });
    }

    return c.json({ success: true, message: `Issue ${issueId} reopened` });
  } catch (error) {
    if (error instanceof HTTPException) throw error;
    const err = error as Error;
    throw new HTTPException(500, { message: err.message });
  }
});

// POST /api/projects/:name/beads/sync - Sync with git remote
beadsApiRouter.post('/:name/beads/sync', async (c) => {
  const projectName = validateProjectName(c.req.param('name'));

  try {
    const manager = await getBeadsManager(projectName);
    const [success, message] = await manager.sync();

    if (!success) {
      throw new HTTPException(500, { message });
    }

    return c.json({ success: true, message: 'Beads synced with remote' });
  } catch (error) {
    if (error instanceof HTTPException) throw error;
    const err = error as Error;
    throw new HTTPException(500, { message: err.message });
  }
});

// POST /api/projects/:name/beads/dep/add - Add a dependency
beadsApiRouter.post('/:name/beads/dep/add', async (c) => {
  const projectName = validateProjectName(c.req.param('name'));
  const body = await parseBody(c, DependencyAddSchema);

  const issueId = validateIssueId(body.issue_id);
  const dependsOn = validateIssueId(body.depends_on);

  try {
    const manager = await getBeadsManager(projectName);
    const result = await manager.addDependency(issueId, dependsOn);

    if (result.error) {
      throw new HTTPException(500, { message: result.error });
    }

    return c.json({ success: true, message: `Added dependency: ${issueId} depends on ${dependsOn}` });
  } catch (error) {
    if (error instanceof HTTPException) throw error;
    const err = error as Error;
    throw new HTTPException(500, { message: err.message });
  }
});

// POST /api/projects/:name/beads/comments/:issueId - Add a comment
beadsApiRouter.post('/:name/beads/comments/:issueId', async (c) => {
  const projectName = validateProjectName(c.req.param('name'));
  const issueId = validateIssueId(c.req.param('issueId'));
  const body = await parseBody(c, CommentAddSchema);

  try {
    const manager = await getBeadsManager(projectName);
    const result = await manager.addComment(issueId, body.comment);

    if (result.error) {
      if (result.error.toLowerCase().includes('not found')) {
        throw new HTTPException(404, { message: `Issue ${issueId} not found` });
      }
      throw new HTTPException(500, { message: result.error });
    }

    return c.json({ success: true, message: `Comment added to ${issueId}` });
  } catch (error) {
    if (error instanceof HTTPException) throw error;
    const err = error as Error;
    throw new HTTPException(500, { message: err.message });
  }
});

export { beadsApiRouter };
